BlueprintUICommonControls/Sources/Internal/KeyboardObserver.swift

```diff
@@ -0,0 +1,178 @@
+//
+//  KeyboardObserver.swift
+//  BlueprintUICommonControls
+//
+//  Created by Kyle Van Essen on 2/16/20.
+//
+
+import UIKit
+
+
+protocol KeyboardObserverDelegate : AnyObject {
+    
+    func keyboardFrameWillChange(
+        for observer : KeyboardObserver,
+        animationDuration : Double,
+        options : UIView.AnimationOptions
+    )
+}
+
+/**
+ Implementation borrowed from Listable:
+ https://github.com/kyleve/Listable/blob/master/Listable/Sources/Internal/KeyboardObserver.swift
+ 
+ iOS Docs for keyboard management:
+ https://developer.apple.com/library/archive/documentation/StringsTextFonts/Conceptual/TextAndWebiPhoneOS/KeyboardManagement/KeyboardManagement.html
+ */
+final class KeyboardObserver {
+    
+    private let center : NotificationCenter
+    
+    weak var delegate : KeyboardObserverDelegate?
+    
+    //
+    // MARK: Initialization
+    //
+    
+    init(center : NotificationCenter = .default) {
+        
+        self.center = center
+        
+        /// We need to listen to both `will` and `keyboardDidChangeFrame` notifications. Why?
+        /// When dealing with an undocked or floating keyboard, moving the keyboard
+        /// around the screen does NOT call `willChangeFrame`; only `didChangeFrame` is called.
+        /// Before calling the delegate, we compare `old.endingFrame != new.endingFrame`,
+        /// which ensures that the delegate is notified if the frame really changes, and
+        /// prevents duplicate calls.
+
+        self.center.addObserver(self, selector: #selector(keyboardFrameChanged(_:)), name: UIWindow.keyboardWillChangeFrameNotification, object: nil)
+        self.center.addObserver(self, selector: #selector(keyboardFrameChanged(_:)), name: UIWindow.keyboardDidChangeFrameNotification, object: nil)
+    }
+    
+    private var latestNotification : NotificationInfo?
+    
+    //
+    // MARK: Handling Changes
+    //
+    
+    enum KeyboardFrame : Equatable {
+        
+        /// The current frame does not overlap the current view at all.
+        case nonOverlapping
+        
+        /// The current frame does overlap the view, by the provided rect, in the view's coordinate space.
+        case visible(frame: CGRect)
```

Comment:
Should this case be named `overlapping` to match the other case?

Reply:
👍🏼